en/middleware/mavlink.md

```diff
@@ -2,200 +2,419 @@
+This is the kind of implementation that you would provide to support a MAVLink battery integration with PX4.
 
-```C
-#include <uORB/topics/ca_trajectory.h>
-#include <v2.0/custom_messages/mavlink_msg_ca_trajectory.h>
+Add the headers for the incoming MAVLink message and the uORB topic to publish to in [mavlink_receiver.h](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.h#L77):
+
+```cpp
+#include <uORB/topics/battery_status.h>
+#include <v2.0/development/mavlink.h>
 ```
 
-Add a function that handles the incoming MAVLink message in the `MavlinkReceiver` class in
-[mavlink_receiver.h](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.h#L140)
+Add a function signature for a function that handles the incoming MAVLink message in the `MavlinkReceiver` class in
+[mavlink_receiver.h](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.h#L126)
 
-```C
-void handle_message_ca_trajectory_msg(mavlink_message_t *msg);
+```cpp
+void handle_message_battery_status_demo(mavlink_message_t *msg);
 ```
-Add an uORB publisher in the `MavlinkReceiver` class in
-[mavlink_receiver.h](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.h#L195)
 
-```C
-uORB::Publication<ca_trajectory_s>			_ca_traj_msg_pub{ORB_ID(ca_trajectory)};
+Normally you would add a uORB publisher for the uORB topic to publish in the `MavlinkReceiver` class in
+[mavlink_receiver.h](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.h#L296).
+In this case the [BatteryStatus](../msg_docs/BatteryStatus.md) uORB topic already exists:
+
+```cpp
+uORB::Publication<battery_status_s> _battery_pub{ORB_ID(battery_status)};
 ```
 
-Implement the `handle_message_ca_trajectory_msg` function in [mavlink_receiver.cpp](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.cpp)
+Implement the `handle_message_battery_status_demo` function in [mavlink_receiver.cpp](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.cpp).
 
-```C
-void MavlinkReceiver::handle_message_ca_trajectory_msg(mavlink_message_t *msg)
+```cpp
+void
+MavlinkReceiver::handle_message_battery_status_demo(mavlink_message_t *msg)
 {
-    mavlink_ca_trajectory_t traj;
-    mavlink_msg_ca_trajectory_decode(msg, &traj);
-
-    struct ca_traj_struct_s f;
-    memset(&f, 0, sizeof(f));
-
-    f.timestamp = hrt_absolute_time();
-    f.seq_id = traj.seq_id;
-    f.time_start_usec = traj.time_start_usec;
-    f.time_stop_usec = traj.time_stop_usec;
-    for(int i=0;i<28;i++)
-        f.coefficients[i] = traj.coefficients[i];
-        
-    _ca_traj_msg_pub.publish(f);
+	if ((msg->sysid != mavlink_system.sysid) || (msg->compid == mavlink_system.compid)) {
```

Comment:
So what instance are we writing to? It looks like maybe "the default instance"? What if there are two mavlink batteries and one "internal" one? How do we sort that out? I don't plan to say too much but we should comment on what is happening.

Reply:
`uORB::Publication<battery_status_s>` means to use single-instance. For multiple it really depends on what you want to achieve. The simplest (and automatic) is to use an array of `uORB::PublicationMulti`.

Reply:
@bkueng I don't think that quite answers my question. For the incoming battery status message (the current/real one) we have this single publication, so what we have here is a single topic.

```cpp
uORB::Publication<battery_status_s>	_battery_pub{ORB_ID(battery_status)};
```

But when we're streaming we have multiple instances we iterate through and publish a MAVLink message for each.

My inference is that there are multiple battery instance definitions possible in PX4 but we only write into to one of them - if that is correct (?) then which one - the first item, which I expect is the battery with mavlink instance id `1`?

Further doesn't this imply that while we can emit MAVLink messages for any number of connected batteries, we can only ever have one "incoming" battery connected to PX4 via MAVLink?

Lastly, given your answer, I think that this means that if we wanted to have multiple batteries "incoming" and connected via MAVLink we'd need to use a multiple publication, and take the instance number to work out which one to write to. Right?

I'm not going to block on this but if you answer I'll put in an update on my return.

----

As an aside, I think we/I need to go through uORB docs again and document this both as a guide and as a reference - explaining all these methods for subscription. The current docs are ... insufficient.

Reply:
> My inference is that there are multiple battery instance definitions possible in PX4 but we only write into to one of them - if that is correct (?) then which one - the first item, which I expect is the battery with mavlink instance id 1?

Yes in that case it would publish to the first uorb instance. The battery case is a bit complex for that reason. The message also has an `id` field that's used for mavlink.

> Lastly, given your answer, I think that this means that if we wanted to have multiple batteries "incoming" and connected via MAVLink we'd need to use a multiple publication, and take the instance number to work out which one to write to. Right?

Yes that's where `PublicationMulti` comes in.

Reply:
Added explanation in https://github.com/PX4/PX4-user_guide/pull/3027